File: frameworks/webonyx-graphql-php/tests/SecurityInjectionTest.php
<?php

declare(strict_types=1);

namespace VelocityBench\Tests;

use PHPUnit\Framework\TestCase;

require_once __DIR__ . '/TestFactory.php';

/**
 * SecurityInjectionTest - Tests SQL injection prevention
 *
 * Coverage includes:
 * - Basic OR injection attempts
 * - UNION-based injection attempts
 * - Stacked queries
 * - Comment sequence injection
 * - Time-based blind injection attempts
 */
class SecurityInjectionTest extends TestCase
{
    private TestFactory $factory;

    protected function setUp(): void
    {
        $this->factory = new TestFactory();
    }

    protected function tearDown(): void
    {
        $this->factory->reset();
    }

    // ============================================================================
    // SQL Injection Prevention Tests
    // ============================================================================

    public function testPreventsBasicSqlInjection(): void
    {
        $user = $this->factory->createUser('alice', 'alice@example.com', 'Alice');
        $maliciousInput = "1' OR '1'='1";

        // Attempt to query with malicious input
        $result = $this->factory->getUser($maliciousInput);

        // Should return null, not all users
        $this->assertNull($result);
    }

    public function testPreventsUnionBasedInjection(): void
    {
        $user = $this->factory->createUser('bob', 'bob@example.com', 'Bob');
        $maliciousInput = "1' UNION SELECT * FROM users--";

        // Attempt injection
        $result = $this->factory->getUser($maliciousInput);

        // Should handle safely
        $this->assertNull($result);
    }

    public function testPreventsStackedQueries(): void
    {
        $user = $this->factory->createUser('charlie', 'charlie@example.com', 'Charlie');
        $maliciousInput = "1'; DROP TABLE users;--";

        // Attempt to execute stacked queries
        $result = $this->factory->getUser($maliciousInput);

        // Should handle safely
        $this->assertNull($result);

        // Verify data still exists (table not dropped)
        $verifyUser = $this->factory->getUser($user->id);
        $this->assertNotNull($verifyUser);
    }

    public function testPreventsCommentSequenceInjection(): void
    {
        $user = $this->factory->createUser('dave', 'dave@example.com', 'Dave');
        $maliciousInput = "admin'--";

        // Attempt comment-based bypass
        $result = $this->factory->getUser($maliciousInput);

        $this->assertNull($result);
    }

    public function testPreventsTimeBasedBlindInjection(): void
    {
        $user = $this->factory->createUser('eve', 'eve@example.com', 'Eve');
        $maliciousInput = "1' AND SLEEP(5)--";

        // Measure execution time
        $startTime = microtime(true);
        $result = $this->factory->getUser($maliciousInput);
        $endTime = microtime(true);
        $duration = $endTime - $startTime;

        // Should not execute SLEEP
        $this->assertNull($result);
        $this->assertLessThan(2.0, $duration, 'Query should not execute SLEEP command');
    }

    public function testPreventsInjectionInPostTitle(): void
    {
        $author = $this->factory->createUser('author', 'author@example.com', 'Author');
        $post = $this->factory->createPost($author->id, 'Test Post', 'Content');
        $maliciousInput = "1' OR '1'='1";

        // Try to search posts with malicious title
        // Since TestFactory doesn't have search, we test safe retrieval
        $result = $this->factory->getPost($maliciousInput);

        $this->assertNull($result);
    }

    public function testHandlesSpecialCharactersSafely(): void
    {
        $author = $this->factory->createUser('author', 'author@example.com', 'Author');
        $specialContent = "Test with 'quotes' and \"double quotes\" and <html> tags";
        $post = $this->factory->createPost($author->id, 'Special', $specialContent);

        // Retrieve and verify special characters are preserved
        $retrieved = $this->factory->getPost($post->id);

        $this->assertNotNull($retrieved);
        $this->assertEquals($specialContent, $retrieved->content);
    }

    public function testPreventsInjectionViaNumericId(): void
    {
        $author = $this->factory->createUser('author', 'author@example.com', 'Author');
        $post = $this->factory->createPost($author->id, 'Test', 'Content');
        $maliciousId = "1 OR 1=1";

        // Should handle non-UUID format safely
        $result = $this->factory->getPost($maliciousId);

        $this->assertNull($result);
    }

    public function testPreventsHexEncodedInjection(): void
    {
        $user = $this->factory->createUser('frank', 'frank@example.com', 'Frank');
        $maliciousInput = "0x61646D696E"; // Hex for 'admin'

        $result = $this->factory->getUser($maliciousInput);

        $this->assertNull($result);
    }

    public function testPreventsInjectionWithMultibyteCharacters(): void
    {
        $user = $this->factory->createUser('user', 'user@example.com', 'User');
        $maliciousInput = "admin' OR '1'='1' /*中文注入*/";

        $result = $this->factory->getUser($maliciousInput);

        $this->assertNull($result);
    }

    public function testValidatesUuidFormatStrict(): void
    {
        $user = $this->factory->createUser('user', 'user@example.com', 'User');
        $invalidUuid = "not-a-uuid-format";

        // Should reject non-UUID format
        $result = $this->factory->getUser($invalidUuid);

        $this->assertNull($result);
    }

    public function testHandlesNullByteInjection(): void
    {
        $user = $this->factory->createUser('user', 'user@example.com', 'User');
        $maliciousInput = "admin\x00'--";

        $result = $this->factory->getUser($maliciousInput);

        $this->assertNull($result);
    }

    public function testPreventsSecondOrderInjection(): void
    {
        // Create user with malicious content
        $maliciousUsername = "admin' OR '1'='1";

        try {
            $user = $this->factory->createUser($maliciousUsername, 'mal@example.com', 'Malicious');

            // If creation succeeds, retrieval should handle safely
            $posts = $this->factory->getPostsByAuthor($user->pkUser);

            // Should return empty array, not all posts
            $this->assertIsArray($posts);
        } catch (\Exception $e) {
            // If creation fails due to validation, that's also acceptable
            $this->assertTrue(true);
        }
    }

    public function testParameterizedQueriesWithUnicode(): void
    {
        $unicodeContent = 'Test with émojis 🎉 and ñ and 中文 and Ελληνικά';
        $author = $this->factory->createUser('author', 'author@example.com', 'Author');
        $post = $this->factory->createPost($author->id, 'Unicode Test', $unicodeContent);

        // Retrieve and verify Unicode is preserved
        $retrieved = $this->factory->getPost($post->id);

        $this->assertNotNull($retrieved);
        $this->assertEquals($unicodeContent, $retrieved->content);
    }

    public function testLongInputDoesNotCauseInjection(): void
    {
        $user = $this->factory->createUser('user', 'user@example.com', 'User');
        $longMaliciousInput = str_repeat("' OR '1'='1", 1000);

        // Should handle long inputs safely
        $result = $this->factory->getUser($longMaliciousInput);

        $this->assertNull($result);
    }
}

File: frameworks/csharp-dotnet/FraiseQL.Benchmark.Tests/SecurityInjectionTest.cs
using FraiseQL.Benchmark.Models;
using Xunit;
using System;
using System.Diagnostics;

namespace FraiseQL.Benchmark.Tests;

/// <summary>
/// SecurityInjectionTest - Tests SQL injection prevention
///
/// Coverage includes:
/// - Basic OR injection attempts
/// - UNION-based injection attempts
/// - Stacked queries
/// - Comment sequence injection
/// - Time-based blind injection attempts
/// </summary>
public class SecurityInjectionTest
{
    private readonly TestFactory _factory = new();

    // ============================================================================
    // SQL Injection Prevention Tests
    // ============================================================================

    [Fact]
    public void PreventBasicSqlInjection()
    {
        var user = _factory.CreateTestUser("alice", "alice@example.com", "Alice", "");
        var maliciousInput = "1' OR '1'='1";

        // Attempt to query with malicious input - invalid format is safely rejected
        User? result = null;
        if (Guid.TryParse(maliciousInput, out var guid))
        {
            result = _factory.GetUser(guid);
        }

        // Should return null, not all users
        Assert.Null(result);
    }

    [Fact]
    public void PreventUnionBasedInjection()
    {
        var user = _factory.CreateTestUser("bob", "bob@example.com", "Bob", "");
        var maliciousInput = "1' UNION SELECT * FROM users--";

        // Should handle invalid GUID format safely
        Assert.Throws<FormatException>(() =>
        {
            var guid = Guid.Parse(maliciousInput);
        });
    }

    [Fact]
    public void PreventStackedQueries()
    {
        var user = _factory.CreateTestUser("charlie", "charlie@example.com", "Charlie", "");
        var maliciousInput = "1'; DROP TABLE users;--";

        // Should handle invalid input safely
        Assert.Throws<FormatException>(() =>
        {
            var guid = Guid.Parse(maliciousInput);
        });

        // Verify data still exists
        var verifyUser = _factory.GetUser(user.Id);
        Assert.NotNull(verifyUser);
    }

    [Fact]
    public void PreventCommentSequenceInjection()
    {
        var user = _factory.CreateTestUser("dave", "dave@example.com", "Dave", "");
        var maliciousInput = "admin'--";

        // Should reject invalid format
        Assert.Throws<FormatException>(() =>
        {
            var guid = Guid.Parse(maliciousInput);
        });
    }

    [Fact]
    public void PreventTimeBasedBlindInjection()
    {
        var user = _factory.CreateTestUser("eve", "eve@example.com", "Eve", "");
        var maliciousInput = "1' AND SLEEP(5)--";

        // Measure execution time
        var stopwatch = Stopwatch.StartNew();

        Assert.Throws<FormatException>(() =>
        {
            var guid = Guid.Parse(maliciousInput);
        });

        stopwatch.Stop();

        // Should not execute SLEEP, response should be fast
        Assert.True(stopwatch.ElapsedMilliseconds < 2000, "Query should not execute SLEEP command");
    }

    [Fact]
    public void PreventInjectionInPostTitle()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var post = _factory.CreateTestPost(author.Id, "Test Post", "Content");
        var maliciousInput = "1' OR '1'='1";

        // Should reject invalid GUID
        Assert.Throws<FormatException>(() =>
        {
            var guid = Guid.Parse(maliciousInput);
        });
    }

    [Fact]
    public void HandleSpecialCharactersSafely()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var specialContent = "Test with 'quotes' and \"double quotes\" and <html> tags";
        var post = _factory.CreateTestPost(author.Id, "Special", specialContent);

        // Retrieve and verify special characters are preserved
        var retrieved = _factory.GetPost(post.Id);

        Assert.NotNull(retrieved);
        Assert.Equal(specialContent, retrieved.Content);
    }

    [Fact]
    public void PreventInjectionViaNonGuidId()
    {
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var post = _factory.CreateTestPost(author.Id, "Test", "Content");
        var maliciousId = "1 OR 1=1";

        // Should handle non-GUID format safely
        Assert.Throws<FormatException>(() =>
        {
            var guid = Guid.Parse(maliciousId);
        });
    }

    [Fact]
    public void PreventHexEncodedInjection()
    {
        var user = _factory.CreateTestUser("frank", "frank@example.com", "Frank", "");
        var maliciousInput = "0x61646D696E"; // Hex for 'admin'

        // Should reject invalid format
        Assert.Throws<FormatException>(() =>
        {
            var guid = Guid.Parse(maliciousInput);
        });
    }

    [Fact]
    public void PreventInjectionWithMultibyteCharacters()
    {
        var user = _factory.CreateTestUser("user", "user@example.com", "User", "");
        var maliciousInput = "admin' OR '1'='1' /*中文注入*/";

        // Should reject invalid format
        Assert.Throws<FormatException>(() =>
        {
            var guid = Guid.Parse(maliciousInput);
        });
    }

    [Fact]
    public void ValidateGuidFormatStrict()
    {
        var user = _factory.CreateTestUser("user", "user@example.com", "User", "");
        var invalidGuid = "not-a-guid-format";

        // Should reject non-GUID format
        Assert.Throws<FormatException>(() =>
        {
            var guid = Guid.Parse(invalidGuid);
        });
    }

    [Fact]
    public void HandleNullByteInjection()
    {
        var user = _factory.CreateTestUser("user", "user@example.com", "User", "");
        var maliciousInput = "admin\0'--";

        // Should reject invalid format
        Assert.Throws<FormatException>(() =>
        {
            var guid = Guid.Parse(maliciousInput);
        });
    }

    [Fact]
    public void PreventSecondOrderInjection()
    {
        // Create user with malicious content in username
        var maliciousUsername = "admin' OR '1'='1";

        // Creation should either succeed (and be escaped) or fail validation
        try
        {
            var user = _factory.CreateTestUser(maliciousUsername, "mal@example.com", "Malicious", "");

            // If creation succeeds, retrieval should handle safely
            var posts = _factory.GetAllPosts()
                .Where(p => p.Author?.Id == user.Id)
                .ToList();

            // Should return empty list or user's posts only
            Assert.NotNull(posts);
        }
        catch (ArgumentException)
        {
            // If validation rejects, that's acceptable
            Assert.True(true);
        }
    }

    [Fact]
    public void ParameterizedQueriesWithUnicode()
    {
        var unicodeContent = "Test with émojis 🎉 and ñ and 中文 and Ελληνικά";
        var author = _factory.CreateTestUser("author", "author@example.com", "Author", "");
        var post = _factory.CreateTestPost(author.Id, "Unicode Test", unicodeContent);

        // Retrieve and verify Unicode is preserved
        var retrieved = _factory.GetPost(post.Id);

        Assert.NotNull(retrieved);
        Assert.Equal(unicodeContent, retrieved.Content);
    }

    [Fact]
    public void LongInputDoesNotCauseInjection()
    {
        var user = _factory.CreateTestUser("user", "user@example.com", "User", "");
        var longMaliciousInput = string.Concat(Enumerable.Repeat("' OR '1'='1", 1000));

        // Should reject invalid format (too long for GUID)
        Assert.Throws<FormatException>(() =>
        {
            var guid = Guid.Parse(longMaliciousInput);
        });
    }

    [Fact]
    public void GuidParsingEnforcesFormat()
    {
        // Valid GUID format
        var validGuid = Guid.NewGuid().ToString();
        var parsed = Guid.Parse(validGuid);
        Assert.NotEqual(Guid.Empty, parsed);

        // Invalid formats should throw
        Assert.Throws<FormatException>(() => Guid.Parse("invalid"));
        Assert.Throws<FormatException>(() => Guid.Parse("1234"));
        Assert.Throws<FormatException>(() => Guid.Parse("' OR '1'='1"));
    }

    [Fact]
    public void SpecialCharactersInNamesPreserved()
    {
        // Names with SQL-like characters should be preserved safely
        var specialName = "O'Brien <script>alert('xss')</script>";
        var user = _factory.CreateTestUser("obrien", "obrien@example.com", specialName, "");

        // Retrieve and verify
        var retrieved = _factory.GetUser(user.Id);

        Assert.NotNull(retrieved);
        Assert.Equal(specialName, retrieved.FullName);
    }
}
